### PR TITLE
[SkyServe] deprecate old serve autoscaler policy

### PR DESCRIPTION
**Pull request by MaoZiming**
After discussing with @cblmemo 
This PR replaces the old Sky Serve autoscaler policy with the new counter-based version (as in the spot policy)

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Manual Test: `sky serve up tests/test_yamls/test_serve_autoscaler.yaml` followed by `python3 tests/test_serve_autoscaler.py`
- [ ] All smoke tests: `pytest tests/test_smoke.py`

### CODE REVIEW

File: sky/serve/autoscalers.py [hunk 1]
```diff
@@ -16,6 +16,9 @@
 
 logger = sky_logging.init_logger(__name__)
 
+_UPSCALE_DELAY_S = 300
```
Comment (cblmemo) [inline sky/serve/autoscalers.py:19]: Should we expose this option to user?

File: sky/serve/autoscalers.py [hunk 2]
```diff
@@ -105,6 +108,12 @@ def __init__(self, spec: 'service_spec.SkyServiceSpec', frequency: int,
         self.rps_window_size: int = rps_window_size
         self.last_scale_operation: float = 0.
```
Comment (cblmemo) [inline sky/serve/autoscalers.py:109]: This variable is not used anymore?

**Review by cblmemo — COMMENTED** [review]
This looks great! Maybe we could migrate more implementation into our on-demand autoscaler, such as:
* Change the `AutoscalerDecision` to the latest version in the spot policy;
* Change the type of return value for `evaluate_scaling` (i.e., `List[AutoscalerDecision]`);
* Using the same interface (`target_qps` instead of `{upper,lower}_threshold`) as spot policy;
  * Probably also migrate the `_get_desired_num_replicas` function.

The latest bullet point needs more discussion; cc @Michaelvll @concretevitamin for a look.

File: sky/serve/autoscalers.py (hunk 1 above)
Comment (MaoZiming) [reply sky/serve/autoscalers.py:19]: current `AUTOSCALER_SCALE_FREQUENCY_SECONDS` is not exposed to the user

**Comment by MaoZiming** [comment]
@cblmemo we probably want to think about how much to migrate from the spot policy PR. The current PR is kinda a minimal version of adding counter-based policy

**Comment by cblmemo** [comment]
Seems like AUTOSCALER_COOLDOWN_SECONDS is deprecated too

**Comment by cblmemo** [comment]
> @cblmemo we probably want to think about how much to migrate from the spot policy PR. The current PR is kinda a minimal version of adding counter-based policy

Yeah, we should discuss more. Agreed that we should migrate the counter-based autoscaling policy to master ASAP

File: sky/serve/autoscalers.py [hunk 3]
```diff
@@ -167,7 +162,24 @@ def evaluate_scaling(
...
+                num_replicas_delta = target_num_replicas - num_replicas
+        elif target_num_replicas < num_replicas:
+            self.downscale_counter += 1
+            self.upscale_counter = 0
+            if self.downscale_counter >= self.scale_down_consecutive_periods:
+                self.downscale_counter = 0
+                num_replicas_delta = target_num_replicas - num_replicas
+        else:
+            self.upscale_counter = self.downscale_counter = 0
+            num_replicas_delta = 0
```
Comment (cblmemo) [inline sky/serve/autoscalers.py:181]: nit: this line is redundant?

File: sky/serve/autoscalers.py [hunk 4]
```diff
@@ -167,7 +162,24 @@ def evaluate_scaling(
 
         target_num_replicas = max(self.min_replicas,
                                   min(self.max_replicas, target_num_replicas))
-        num_replicas_delta = target_num_replicas - num_replicas
+
+        num_replicas_delta = 0
+        if target_num_replicas > num_replicas:
+            self.upscale_counter += 1
+            self.downscale_counter = 0
+            if self.upscale_counter >= self.scale_up_consecutive_periods:
```
Comment (cblmemo) [inline sky/serve/autoscalers.py:170]: Should we log these counters to the user?

**Review by cblmemo — APPROVED** [review]
If we are doing minimal migration then this version of the code looks great to me 🫡

**Comment by MaoZiming** [comment]
> AUTOSCALER_COOLDOWN_SECONDS

https://github.com/skypilot-org/skypilot/blob/serve-spot/sky/serve/constants.py#L41
it is still hard-coded

**Comment by cblmemo** [comment]
> > AUTOSCALER_COOLDOWN_SECONDS
> 
> https://github.com/skypilot-org/skypilot/blob/serve-spot/sky/serve/constants.py#L41 it is still hard-coded

I mean all variable/constants related to cooldown should be able to remove, including this constant and `self.cooldown` for `RequestRateAutoscaler`

**Review by cblmemo — CHANGES_REQUESTED** [review]
Oops, I missed one thing. We should not let the user wait for 5 minutes for the bootstrap case and we need some special judgment for this case.

File: sky/serve/autoscalers.py [hunk 5]
```diff
@@ -167,7 +162,31 @@ def evaluate_scaling(
...
+                num_replicas_delta = target_num_replicas - num_replicas
+        elif target_num_replicas < num_replicas:
+            self.downscale_counter += 1
+            self.upscale_counter = 0
+            if self.downscale_counter >= self.scale_down_consecutive_periods:
+                self.downscale_counter = 0
+                num_replicas_delta = target_num_replicas - num_replicas
+        else:
+            self.upscale_counter = self.downscale_counter = 0
+        logger.info(f'Upscale counter: {self.upscale_counter}. '
```
Comment (cblmemo) [inline sky/serve/autoscalers.py:187]: Should we log the counter threshold? i.e.

```bash
Upscale counter: 9/15
```

File: sky/serve/autoscalers.py [hunk 6]
```diff
@@ -103,8 +106,13 @@ def __init__(self, spec: 'service_spec.SkyServiceSpec', frequency: int,
         self.lower_threshold: Optional[float] = spec.qps_lower_threshold
         self.cooldown: int = cooldown
```
Comment (cblmemo) [inline sky/serve/autoscalers.py:107]: remove these unused constant/variable related to cooldown?

**Review by cblmemo — APPROVED** [review]
It looks great to me! Left some nits, and please run some manual tests and include them in the PR description.

**Review by cblmemo — APPROVED** [review]
There is still a constant to delete IIRC, other than that it looks great to me! After some tests this PR should be ready to go 🫡

**Comment by MaoZiming** [comment]
@cblmemo added a manual test. commented out the constant

File: examples/serve/autoscaler.py
Comment (cblmemo) [inline examples/serve/autoscaler.py:1]: do we really need this..?

File: examples/serve/autoscaler.yaml
Comment (cblmemo) [inline examples/serve/autoscaler.yaml:1]: ditto

**Review by cblmemo — COMMENTED** [review]
sorry for not making it clear.. I mean test to make sure it doesn't contain any unexpected errors. Probably we could eliminate it in the PR since it is very similar to our http server, and the interface will change soon (to target qps per replica).

File: examples/serve/autoscaler.yaml
Comment (MaoZiming) [reply examples/serve/autoscaler.yaml:1]: these are some manual tests for the autoscaling behavior.
Are there other scripts where we can test the skyserve autoscaler?

Comment (cblmemo) [reply examples/serve/autoscaler.yaml:1]: I usually just use watch -n <some interval> curl http://...

Comment (Michaelvll) [reply examples/serve/autoscaler.yaml:1]: For the yamls that are mainly for testing, we can probably have them in `./tests/test_yamls`?

Comment (MaoZiming) [reply examples/serve/autoscaler.yaml:1]: moved to test

**Pull request merged**